otx/hpo/hpo_runner.py

```diff
@@ -0,0 +1,386 @@
+
+    Args:
+        hpo_algo (HpoBase): HPO algorithms.
+        train_func (Callable): Function to train a model.
+        resource_type (Literal['gpu', 'cpu'], optional): Which type of resource to use.
+                                                         If can be changed depending on environment. Defaults to "gpu".
+        num_parallel_trial (Optional[int], optional): How many trials to trun parallelly.
+                                                    It's used for CPUResourceManager. Defaults to None.
+        num_gpu_for_single_trial (Optional[int], optional): How many GPUs is used for a single trial.
+                                                            It's used for GPUResourceManager. Defaults to None.
+        available_gpu (Optional[str], optional): How many GPUs are available. It's used for GPUResourceManager.
+                                                 Defaults to None.
+    """
+
+    def __init__(
+        self,
+        hpo_algo: HpoBase,
+        train_func: Callable,
+        resource_type: Literal["gpu", "cpu"] = "gpu",
+        num_parallel_trial: Optional[int] = None,
+        num_gpu_for_single_trial: Optional[int] = None,
+        available_gpu: Optional[str] = None,
+    ):
+        self._hpo_algo = hpo_algo
+        self._train_func = train_func
+        self._running_trials: Dict[int, Dict] = {}
+        self._mp = multiprocessing.get_context("spawn")
+        self._uid_index = 0
+        self._resource_manager = get_resource_manager(
+            resource_type, num_parallel_trial, num_gpu_for_single_trial, available_gpu
+        )
+
+    def run(self):
+        """Run a HPO loop."""
+        logger.info("HPO loop starts.")
+        while not self._hpo_algo.is_done():
+            if self._resource_manager.have_available_resource():
+                trial = self._hpo_algo.get_next_sample()
+                if trial is not None:
+                    self._start_trial_process(trial)
+
+            self._remove_finished_process()
+            self._get_reports()
+
+        logger.info("HPO loop is done.")
+        self._get_reports()
+        self._join_all_processes()
+
+    def _start_trial_process(self, trial: Trial):
+        logger.info(f"{trial.id} trial is now running.")
+        logger.debug(f"{trial.id} hyper paramter => {trial.configuration}")
+
+        trial.status = TrialStatus.RUNNING
+        uid = self._get_uid()
+
+        origin_env = os.environ
+        env = self._resource_manager.reserve_resource(uid)
+        if env is not None:
+            for key, val in env.items():
+                os.environ[key] = val
+
+        pipe1, pipe2 = self._mp.Pipe(True)
+        process = self._mp.Process(
+            target=_run_train,
+            args=(
+                self._train_func,
+                trial.get_train_configuration(),
+                partial(_report_score, pipe=pipe2, trial_id=trial.id),
+            ),
+        )
+        os.environ = origin_env
```

Comment:
It's hard for me to understand `origin_env = os.environ` and `os.environ = origin_env`. Could you explain for this? Thank you!

Reply:
Purpose of the code is passing appropriate environment variables to child process. But I think I forget to deepcopy the `os.environ`. Thanks for pointing it!